nursery/curses/setup.py

```diff
@@ -1,50 +1,25 @@
-#/usr/bin/env python
-import io
+#!/usr/bin/env python
 import re
 
-from setuptools import setup, find_packages
+from setuptools import setup
 
-with io.open('toga_curses/__init__.py', encoding='utf8') as version_file:
-    version_match = re.search(r"^__version__ = ['\"]([^'\"]*)['\"]", version_file.read(), re.M)
+# Version handline needs to be programatic because
+# we can't import toga_curses to compute the version;
+# and to support versioned subpackage dependencies
+with open('toga_curses/__init__.py', encoding='utf8') as version_file:
```

Comment:
You can get rid of this piece of code if you choose to use [bump2version](https://github.com/c4urself/bump2version).
Since you already touch this piece of code, maybe we should consider add the use of it in this PR or in a future PR.

Using *bump2version* will let you update  *versions* in both files (setup.py and \_\_init\_\_.py) simultaneously without checking that they are similar

Reply:
I've seen bump2version (and other tools that do similar things); but I have to say I'm not very enthusiastic about them.

I agree that this boilerplate "open and read file" logic in setup.py is a little cumbersome. However, you'll notice that *briefcase* doesn't need this logic. It uses the `attr: ` directive in setup.cfg to extract the version number. This isn't (currently) an option for Toga because `toga.__init_.py` has side effects when it is imported. The fact that it doesn't work with toga's use case [is a bug that they fixed 4 days ago](https://github.com/pypa/setuptools/issues/1960) in pypa/setuptools#1753. 

Of course, that won't help us resolve the injection of version-tags into `install_requires` dependencies; however, those are really only needed because we're using development releases. Once we publish 0.3.0 final, we should be able to drop those and migrate to a "full" setup.cfg configuration.

So - why use `attr:` over `bump2version`? For me, it comes down to the source of truth. To my mind, the *code* should be the source of truth for the version number, because that's what matters at runtime. bump2version (and similar) approaches use the configuration file as a source of truth. To me, this is the wrong way around. There should be a single source of truth, and it should be as close as possible to the version that is going to be used at runtime - and that means the `__version__` specifier in the package itself.

bump2version uses it's own configuration file as the source of truth - and then *copies* the version into all the places it is needed. However, by using a copy-based approach, it's (a) prone to error (because the copy might fail, or you might miss a place where the version is needed); and (b) it opens up the possibility of user error, because a human could manually modify one of the places where the version number is used. I agree this is an unlikely problem in practice, but the fact that it is possible *at all* is an indication that there is something wrong in the design.

[setuptools_scm](https://pypi.org/project/setuptools-scm/) takes a slightly better approach in that it uses packaging (and version control system) metadata to determine a version at runtime; so, there is a single source of truth that is shared between the configuration *and* runtime (that single source of truth happens to be defined in the configuration file). If we were going to move to an automated tool (and I'm not 100% sure it's worth it), I'd be more inclined to go with something like `setuptools_scm`, rather than bump2version (which is effectively not much more than a `sed` replacement, IMHO) - but I'm still not convinced it's a better approach over the much simpler `attr:` approach (even if we do need a workaround in the short term).

Reply:
FWIW, I am personally -0 on storing the version number in a place that Python code can read at runtime. If we didn't need to store the version number in a place that Python code can read at runtime, then I think life is simpler, and I think the simpler life has no downsides.

Reply:
@paulproteus I'm confused - are you saying that packages shouldn't have a `__version__` specifier *at all*?

Reply:
Yes :D